library/cwm/src/lib/cwm/multi_status_selector.rb

```diff
@@ -0,0 +1,470 @@
+  #
+  #       def contents
+  #         VBox(
+  #           VWeight(60, super),
+  #           VWeight(40, details)
+  #         )
+  #       end
+  #
+  #       def toggle(item)
+  #         item.toggle
+  #         select_dependencies
+  #         label_event_handler(item)
+  #       end
+  #
+  #       private
+  #
+  #       attr_accessor :products
+  #
+  #       def details
+  #         @details ||= CWM::RichText.new
+  #       end
+  #
+  #       def select_dependencies
+  #         # logic to recalculate auto selected items
+  #       end
+  #
+  #       def label_event_handler(item)
+  #         details.value = item.description
+  #       end
+  #
+  #       class Item < CWM::MultiStateSelector::Item
+  #         attr_reader :status, :dependencies
+  #
+  #         def initialize(product)
+  #           @product = product
+  #           @status = product.status || UNSELECTED
+  #           @dependencies = product.dependencies || []
+  #         end
+  #
+  #         def id
+  #           product.id
+  #         end
+  #
+  #         def label
+  #           product.friendly_name || product.name
+  #         end
+  #
+  #         def description
+  #           # build the item description
+  #         end
+  #
+  #         private
+  #
+  #         attr_reader :product
+  #       end
+  #     end
+  class MultiStatusSelector < CustomWidget
+    # @!method items
+    #   The items collection
+    #   @return [Array<Item>] the collection of available items
+    abstract_method :items
+
+    # @macro seeAbstractWidget
+    def init
+      refresh
+    end
+
+    # @macro seeAbstractWidget
+    def contents
+      HBox(content)
+    end
+
+    # @macro seeAbstractWidget
+    def handle(event)
+      if event["ID"].to_s.match?(Item.event_id)
+        id, fired_by = event["ID"].split(Item.event_id)
+
+        item = find_item(id)
+
+        return unless item
+
+        send("#{fired_by}_event_handler", item)
+      end
+
+      nil
+    end
+
+    # Toggles the status of given item
+    #
+    # Redefine it if needed to perform additional actions before or after toggling the item, like
+    # calculating dependencies for auto selection.
+    #
+    # @param item [Item] item to toggle the status
+    def toggle(item)
+      item.toggle
+    end
+
+  private
+
+    # @macro seeAbstractWidget
+    def handle_all_events
+      true
+    end
+
+    # Handles the event fired by the Item check box input
+    #
+    # @param item [Item] the item that fired the event
+    def input_event_handler(item)
+      toggle(item)
+      refresh
+    end
+
+    # Handles the event fired by the Item check box label
+    #
+    # @param [Item] the item that fired the event
+    def label_event_handler(item)
+      log.debug("Unhandled label event fired by #{item.inspect}")
+    end
+
+    # Returns the item with given id
+    #
+    # @param id [String] the id of the searched item
+    # @return [Item, nil] the item matching with given id
+    def find_item(id)
```

Comment:
as id can be any object with `#to_s` why not allow to pass here also object with to_s ?

Reply:
I changed it, but I'm not sure if

* is what you mean by allowing to pass here an object responding to `#to_s`
* the resulting code is good/simple enough

What do you think now?

Reply:
it looks fine for me.